### ISSUE TITLE
Ability to use a single interceptor for both unary and streaming RPCs

### ISSUE THREAD

**Issue by jhump** [grpc/grpc-go#1805]
After #1801 is completed, it becomes possible for a stream interceptor to be used when processing unary RPCs.

This issue is a feature request for some new API that allows for exactly this. The existing API, for backwards compatibility, will continue to register a stream interceptor that is only used for streaming RPCs. But a new method/dial option could be added that allows for registering a stream interceptor that is used for unary RPCs and/or for both streaming and unary RPCs.

**Comment by dfawley** [comment]
I'd like to redesign our interceptor API.  Please add your list of requirements/nice-to-haves in this bug and I'll compile them here:

Requirements:
- Single interceptor type for unary and streaming
  - If it's difficult to implement, provide an optional convenience wrapper to generate a stream-style interceptor from a single function (like how unary interceptors work today).
- Include access to `ServiceDesc`/`FileDescriptorProto` (#1526)
- Chainable / able to easily register multiple handlers

Nice to Haves:
- Single API for stats handlers and interceptors
- Global registration pattern
- Able to specify via `CallOption`s
- Same/similar type signature as RPC handler itself (Java does this)

This is currently just a quick list based on my memory at the moment.  I intend to add to this as I run into more things.

**Comment by devpro108** [comment]
@dfawley wondering any news on new interceptor design, when it will be available, looking forward to single interceptor for both unary and streams..

**Comment by devpro108** [comment]
Thank you @dfawley

**Comment by jhump** [comment]
@dfawley, I think stale bot got it wrong on this one, too. I see this is "on hold", but probably should not be closed -- AFAICT, it is not pending any further input from the reporter(s).

**Comment by dfawley** [comment]
Yeah, it hit a lot of issues this morning.  It doesn't seem to be listening to the "onlyLabels" setting (or we set it wrong).
